File: .NET/VS2010TrainingKit/Labs/IntroToWF/Source/Ex10-HostedDesigner/End/C#/HelloWorkflow.Activities.Tests/PrePostEmptyBody.cs
// ----------------------------------------------------------------------------------
// Microsoft Developer & Platform Evangelism
// 
// Copyright (c) Microsoft Corporation. All rights reserved.
// 
// THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
// EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES 
// OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
// ----------------------------------------------------------------------------------
// The example companies, organizations, products, domain names,
// e-mail addresses, logos, people, places, and events depicted
// herein are fictitious.  No association with any real company,
// organization, product, domain name, email address, logo, person,
// places, or events is intended or should be inferred.
// ----------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Activities;
using HelloWorkflow.Activities;
using System.Activities.Statements;

namespace HelloWorkflow.Activities.Tests
{

    public sealed class PrePostEmptyBody : Activity<string>
    {
        public const string PreMsgText = "Pre Activity Called Ordinal:{0}";
        public const string PostMsgText = "Post Activity Called  Ordinal:{0}";

        public OutArgument<string> PreMsg { get; set; }
        public OutArgument<string> PostMsg { get; set; }

        public PrePostEmptyBody()
        {
            base.Implementation = new Func<Activity>(CreateBody);
        }

        Activity CreateBody()
        {
            var Ordinal = new Variable<int> { Default = 1 };


            return new PrePostSequence()
            {
                Variables = { Ordinal },
                Pre = new Assign()
                {
                    To = new OutArgument<string>(env => this.PreMsg.Get(env)),
                    Value = new InArgument<string>((env) => string.Format(PreMsgText, Ordinal.Get(env)))
                },
                Post = new Sequence()
                {
                    Activities = {
                        new Assign()
                        {
                            // Increment the count
                            To = new OutArgument<int>(Ordinal),
                            Value = new InArgument<int>((env) => Ordinal.Get(env)+1) 
                        },
                        new Assign()
                        {
                            To = new OutArgument<string>(env => this.PostMsg.Get(env)),
                            Value = new InArgument<string>((env) => string.Format(PostMsgText, Ordinal.Get(env))) 
                        }
                    }
                }
            };
        }
    }
}

File: .NET/VS2010TrainingKit/Labs/IntroToWF/Source/Ex10-HostedDesigner/Begin/VB/HelloWorkflow.Activities.Tests/PrePostEmptyBody.vb
' ----------------------------------------------------------------------------------
' Microsoft Developer & Platform Evangelism
' 
' Copyright (c) Microsoft Corporation. All rights reserved.
' 
' THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES 
' OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
' ----------------------------------------------------------------------------------
' The example companies, organizations, products, domain names,
' e-mail addresses, logos, people, places, and events depicted
' herein are fictitious.  No association with any real company,
' organization, product, domain name, email address, logo, person,
' places, or events is intended or should be inferred.
' ----------------------------------------------------------------------------------

Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Activities
Imports HelloWorkflow.Activities
Imports System.Activities.Statements


Public NotInheritable Class PrePostEmptyBody
    Inherits Activity(Of String)
    Public Const PreMsgText As String = "Pre Activity Called Ordinal:{0}"
    Public Const PostMsgText As String = "Post Activity Called  Ordinal:{0}"

    Public Property PreMsg() As OutArgument(Of String)
        Get
            Return m_PreMsg
        End Get
        Set(ByVal value As OutArgument(Of String))
            m_PreMsg = value
        End Set
    End Property
    Private m_PreMsg As OutArgument(Of String)
    Public Property PostMsg() As OutArgument(Of String)
        Get
            Return m_PostMsg
        End Get
        Set(ByVal value As OutArgument(Of String))
            m_PostMsg = value
        End Set
    End Property
    Private m_PostMsg As OutArgument(Of String)

    Public Sub New()
        MyBase.Implementation = New Func(Of Activity)(AddressOf CreateBody)
    End Sub

    Private Function CreateBody() As Activity
        Dim pps As New PrePostSequence

        Dim Ordinal = New Variable(Of Integer)() With {.[Default] = 1}
        pps.Variables.Add(Ordinal)

        pps.Pre = New Assign() With {
          .[To] = New OutArgument(Of String)(Function(env) Me.PreMsg.[Get](env)),
          .Value = New InArgument(Of String)(Function(env) String.Format(PreMsgText, Ordinal.[Get](env)))
         }

        Dim postSequence = New Sequence()

        postSequence.Activities.Add(New Assign() With {
           .[To] = New OutArgument(Of Integer)(Ordinal),
           .Value = New InArgument(Of Integer)(Function(env) Ordinal.[Get](env) + 1)
          })

        postSequence.Activities.Add(New Assign() With {
           .[To] = New OutArgument(Of String)(Function(env) Me.PostMsg.[Get](env)),
           .Value = New InArgument(Of String)(Function(env) String.Format(PostMsgText, Ordinal.[Get](env)))
          })

        pps.Post = postSequence

        Return pps
    End Function
End Class
